[Persistence] Handling of missing persisted projects

Jasu opened the pull request:
Currently, if a previously persisted project does not exist, Treemacs will remove it without prompting, and this change will be persisted when quitting Emacs. This becomes a problem when you happen to start Emacs without your network mounts and the only option for recovery is to `kill -9` Emacs.

This PR adds a new setting `treemacs-missing-project-action` with options `ask`, `remove`, and `keep`. Ask is the default, since deleting user's configuration without prompting is not very nice, and `keep` might leave users thinking that Treemacs is broken because a project suddenly refuses open.

Treemacs seems to handle missing directories pretty well, so I didn't add any state about existence to the project nodes.

The `ask` option will give an abort/retry/fail -style prompt, with options to keep, remove, or try again.

Additionally, `treemacs--restore` temporarily removes `treemacs--persist` from `emacs-kill-hook`, otherwise closing the Emacs window during the `ask` prompt would obliterate any persisted state. I'm not sure if this fixes other issues as well, i.e. if it was previously possible to kill Emacs during `treemacs--restore`.

Alexander-Miller commented:
LGTM, you just need to add the new option to the readme, both the Configuration and Installation sections.

Jasu commented:
Added to docs.

Jasu commented:
I made a new fix on top of this, regarding Tramp - currently, if a project is behind a Tramp connection, `file-exists-p` signals an error instead of returning `nil`, causing persisted workspace to be lost.

Should we merge this first, and then I could open a new PR? Or should I add the change as a new commit to this PR?

Alexander-Miller commented:
> file-exists-p signals an error instead of returning nil

Well that's convenient.

> Or should I add the change as a new commit to this PR?

Let's go with this. Since this is all about fixes  and additions to persisted workspaces it should come as a single commit.

Jasu commented:
I rebased this on the current master, squashed the doc commit, added Tramp compatibility handling:

- Add a macro `treemacs-with-demoted-errors`, which in addition to demoting errors prevents Tramp from popping up the connection window.
- Add a macro `treemacs-with-ignored-errors`, which ignores specific errors.
- Ignore "No file notification program found" notify-file-error when starting a file watch. This happens when expanding a directory with a Tramp connection that does not support file notifications.
- Demote errors in file-readable-p when updating the default directory.
- Demote errors in file-exists-p when loading the workspace.
- Demote errors in file-symlink-p when rendering a project.
- Don't check for file existence of remote project paths when editing a workspace.

Alexander-Miller commented:
> Demote errors in file-exists-p when loading the workspace.

Might not be necessary. Please check whether `tramp-handle-file-symlink-p` and `tramp-handle-file-exists-p` work with remote files.(I would try myself my I've have zero experience with tramp and no remote files to connect to either) If that works then we can just copy and inline those functions.

Jasu commented:
I'm not sure I got what you meant, but calling `tramp-handle-file-exists-p` on a non-connectable remote file attempts to open the connection and fails.

One alternative would be to check whether the connection is already open with e.g. `(or (not (file-remote-p path)) (file-remote-p path nil t))` and only then do file operations on the paths. This might be better than ignoring errors, since Treemacs wouldn't open remote connection without explicit request.

Jasu commented:
... Come to think of it, opening the connection only when requested is the only sane option. I'm using Tramp only to access local Docker containers where it doesn't really matter if it opens a connection. With SSH connections there would be security and usability issues. I was focusing on my use case so this didn't occur to me earlier.

Alexander-Miller commented:
> I'm not sure I got what you meant, but calling tramp-handle-file-exists-p on a non-connectable remote file attempts to open the connection and fails.

I was hoping the tramp versions of these functions would work just as well on local and remote files, that they could determine if a file exists without blowing up in either case and needing to have their errors caught.

> Come to think of it, opening the connection only when requested is the only sane option.

All right, let's use those checks then.

Jasu commented:
I noticed that `find-file-node` sometimes triggers the Tramp error in a call to `f-slash`, as `f-slash` calls `is-file-directory-p`.

Do you think that adding a text property to root-level nodes like `:path-status` or something (I couldn't immediately come up with a good name) would be a good idea? It could have values like `'path-accessible`, `'remote-disconnected`, and `'not-fs-path` for extensions. This way, accidentally opening the connection when scanning for nodes could be avoided by checking that first. Or maybe assigning discconnected projects a non-string path, and using that globally as a criterion for paths you shouldn't be performing FS ops on? It should be sufficient to add this for root nodes, so it would not result in a huge amount of text properties.

This way, every FS call wouldn't need to be wrapped in checks for remote files, since obviously directories cannot contain Tramp paths as children.

The field cannot easily be added to the project `defstruct`, since `find-project-for-node` is what triggers the error when scanning for a file.

Alexander-Miller commented:
Adding such a field to the project is my preferred solution. Conceptually I think this is where such "high-level" information belongs, while text properties mostly story "low-level" display-related information. In my estimation it would also require the least amount of refactoring, just add the field, set it when the project is created, and check when necessary.

Changing the `:path` would mean adjusting and probably `pcase`ing a bunch of access points. A `:path-status` could also work, but here we'd have to pay attention when rendering projects.

> The field cannot easily be added to the project defstruct, since find-project-for-node is what triggers the error when scanning for a file.

That f-slash call is quite immaterial, it's only there to make sure /x/abc becomes /x/abc/ so it won't be seen as a child of /x/abcd. It's trivial to do manually instead. That would solve triggering the error in `find-project-for-node`. 

Do you see any other cases where the path status being saved in the struct would lead to issues? If not then that's where it should be saved. If there's non-trivial obstacles then the text-property solution is a good second choice.

> I couldn't immediately come up with a good name

Neither can I, so let's roll with what we have.

And now that I think about it another advantage of keeping the path-status at hand is using it to decide on colors: projects that were missing when loaded or cannot be accessed remotely can be displayed in a nice grey, maybe even with a different icon.

Jasu commented:
> Do you see any other cases where the path status being saved in the struct would lead to issues? If not then that's where it should be saved. If there's non-trivial obstacles then the text-property solution is a good second choice.

Nope, that was the only one. I'll go with the project struct.

> And now that I think about it another advantage of keeping the path-status at hand is using it to decide on colors: projects that were missing when loaded or cannot be accessed remotely can be displayed in a nice grey, maybe even with a different icon.

And maybe (later, not in this PR) highlight disconnected Tramp projects differently. Tramp doesn't have a hook for connection changes though, so that would likely have to happen in a timer, or by advising some Tramp functions. So it would have an ugly solution if implemented.

Alexander-Miller commented:
> I'll be busy for a while, so I probably won't be touching this until weekend.

Time enough for me to swap all the f-slash calls for my own thing.

Alexander-Miller commented:
The `f-slash` replacement is now pushed. I have also renamed the `treemacs-impl` module to `treemacs-core-utils`. That name was appropriate when it contained all the implementation details while treemacs.el defined the user-facing interface, but treemacs has long grown beyond that point.

Jasu commented:
Nice! I might take a quick look at this PR today, and a less quick one tomorrow.

By the way, what was this about, some "not known to be defined" warnings? [[Misc] Fix warnings about projects-end marker.](https://github.com/Alexander-Miller/treemacs/commit/fa1c80e97e075e59ac628bc73adca23e9b635341)

Could I have noticed the warnings somehow, so I won't make the same mistake again?

Alexander-Miller commented:
> By the way, what was this about, some "not known to be defined" warnings? [Misc] Fix warnings about projects-end marker.

You used the variables from a module that didn't require the one that defined it (couldn't require it either or we would've had a circle) so I added a getter. Flycheck should have warned when you were editing workspaces.el, but strangely enough `make lint` didn't see anything wrong, so CI was clean.

Jasu commented:
Thanks, I'll see if there's someting wrong with my Flycheck. Opening that commit gave me no errors.

Jasu commented:
Still working on this, in particular updating the project status when it's expanded etc. Also have to fix the tests. I added faces for different project states, but did not yet add them to docs (or test all of them)

Alexander-Miller commented on src/elisp/treemacs-macros.el:
```diff
@@ -467,6 +467,23 @@ treemacs buffer exists at all, BODY will be executed."
+(defmacro treemacs-with-ignored-errors (ignored-errors &rest body)
+  "Evaluate BODY with specific errors ignored.
+
+IGNORED-ERRORS is a list of errors to ignored.  Each element is a list whose
```
to ignore~d~

Jasu replied:
Fixed

Jasu commented:
Now I at least feel like I'm done - fixed the tests (and added a few), added docs for the faces.

There are two commits not directly related to this PR:
324896be2eef6bcde8afea102c45fa771f49f511 fixes `treemacs-project-of-node` for extensions where the first component of path is `:custom`. I believe that this didn't cause problems before this PR.
f744fbb0aa19facc90f1ae6bbcde37cd45bb5eb0 fixes `treemacs--find-custom-dir-node`, which I committed already broken in a previous PR, and noticed only coincidentally.

Do you want separate PRs of these?

Alexander-Miller commented:
Moving those two elsewhere won't make this PR much smaller, so let's keep them.

Jasu commented:
I found a bug in this, don't merge yet

Jasu commented on src/elisp/treemacs-mode.el:
```diff
@@ -328,9 +328,10 @@ If there is no node at point use \"~/\" instead.
              (path (or (treemacs-button-get btn :default-directory)
                        (treemacs--nearest-path btn))))
-      (when (and (stringp path)
+      (when (and (treemacs-project->is-readable? project)
```
Here's a bug. Checking that the project is readable is not sufficient, since projects may contain project/dir extensions and tag nodes.

Jasu replied:
Fixed this one by making `treemacs--nearest-path` skip extension paths. I also changed the function into a recursive implementation, simplifying it. I believe it won't cause problems, since the number of skipped non-file nodes is likely small (until someone makes a hugely nested extension.)

Jasu commented:
Rebased on `master`.

Jasu commented:
Fixed a bug in Git handling and force-pushed.

Alexander-Miller commented:
Phew, first review is done.

Jasu commented on src/elisp/treemacs-workspaces.el:
```diff
@@ -138,18 +138,26 @@ Will return `point-min' if there is no next project."
   (inline-quote
    (setq treemacs--project-positions (make-hash-table :test #'equal :size 20))))
 
+(define-inline treemacs-project->key (project)
```
This happens with root-level extensions [here](https://github.com/Alexander-Miller/treemacs/blob/master/src/elisp/treemacs-extensions.el#L443) and [here](https://github.com/Alexander-Miller/treemacs/blob/master/src/elisp/treemacs-extensions.el#L465) - it might be considered a bug though...

Jasu commented on src/elisp/treemacs-workspaces.el:
```diff
@@ -138,18 +138,26 @@ Will return `point-min' if there is no next project."
+  "Get the hash table key of PROJECT."
+  (inline-letevals (project)
+    (inline-quote
+     (if (symbolp ,project)
```
This happens with root-level extensions [here](https://github.com/Alexander-Miller/treemacs/blob/master/src/elisp/treemacs-extensions.el#L443) and [here](https://github.com/Alexander-Miller/treemacs/blob/master/src/elisp/treemacs-extensions.el#L465) - it might be considered a bug though...

Jasu replied:
I changed this to `treemacs-project-p`, since `:root-key-form` does not have to be a symbol.

Alexander-Miller replied:
Not a bug. This is an explicitly wanted exception. Using a project object as a key to save its position works most of the time, but variadic extensions really throw a wrench into those gears. If you don't know how many top-level extension nodes you produce and what they will be called then insisting on consistenly using structs as keys is more trouble than it's worth and using the project's key in such cases is IMO the lesser evil.

Jasu replied:
Thanks for the explanation, makes sense (mostly - I'm not super familiar with variadic extensions though).

Alexander-Miller replied:
Neither am I - they are the next big thing I will be tackling. A good use-case is a package like https://github.com/emacs-lsp/lsp-treemacs. With variadic extensions it will be able to display any combination of 0..3 top-level nodes: Infos and/or Warnings and/or Errors.

Updating will be the interesting part, since entire subtrees may disappear or be added to the dom. There's a lengthy discussion in https://github.com/Alexander-Miller/treemacs/issues/331 if you're interested.

Jasu replied:
I'm definitely interested, as I want to update the nodes in `treemacs-bookmark` when bookmarks change.

Alexander-Miller replied:
There are already methods to update and delete single nodes, see `treemacs-update-node` and `treemacs-delete-single-node`. You only need to wait if your extension is variadic (I haven't tried your bookmark extension yet, it's next on my todo list after integrating this).

Jasu commented on src/elisp/treemacs-async.el:
```diff
@@ -293,15 +293,18 @@ FILE: Filepath"
  * The single directories being collapsed, to be put under filewatch
    if `treemacs-filewatch-mode' is on."
-  (when (> treemacs-collapse-dirs 0)
+  (when (and (> treemacs-collapse-dirs 0)
```
I had rebased the branch on master, so no conflict here. Not sure why Github shows this change coming from my branch though.

Jasu commented:
Thanks for the review, all feedback should now be addressed. I renamed the commit adding faces, since it also added the behavior for checking project status when expanding.

Jasu commented on test/test-treemacs.el:
```diff
@@ -67,29 +67,29 @@
-             (p2 (make-treemacs-project :name "P2" :path "~/B"))
-             (p3 (make-treemacs-project :name "P3" :path "~/C"))
+             (p1 (make-treemacs-project :name "P1" :path "~/A" :path-status 'local-readable))
+             (p2 (make-treemacs-project :name "P2" :path "~/B :path-status 'local-readable"))
```
Tests passed even without the `:path-status`es, I just added them to make the tests match reality.

Jasu commented on test/test-treemacs.el:
```diff
@@ -938,6 +938,10 @@
     (it "Succeeds on correctly formed input"
       (-let [lines '("* W1" "** P1" " - path :: a" "** P2" "- path :: b" "* W2" "** P3" " - path :: c")]
+        (expect (treemacs--validate-persist-lines lines) :to-be 'success)))
+    (it "Succeeds with non-connectable remotes"
```
This is fixed, GitHub didn't update the snippet for some reason (at least for me), but it shows with the newline in the `Files changed` tab.

Alexander-Miller commented on src/elisp/treemacs-workspaces.el:
```diff
@@ -138,18 +139,29 @@ Will return `point-min' if there is no next project."
    (setq treemacs--project-positions (make-hash-table :test #'equal :size 20))))
 
+(define-inline treemacs-project->key (self)
+  "Get the hash table key of PROJECT."
```
PROJECT should be SELF as well.

Jasu replied:
This is now fixed - I also added a comment that `self` does not have to be a project struct.

Alexander-Miller commented:
Aaaand we have a lift off.

Jasu commented:
:tada: Awesome, thanks!
